crates/bevy_core_pipeline/src/deferred/node.rs

```diff
@@ -0,0 +1,192 @@
+
+/// Render node used by the prepass.
+///
+/// By default, inserted before the main pass in the render graph.
+#[derive(Default)]
+pub struct DeferredNode;
+
+impl ViewNode for DeferredNode {
+    type ViewQuery = (
+        &'static ExtractedCamera,
+        &'static RenderPhase<Opaque3dDeferred>,
+        &'static RenderPhase<AlphaMask3dDeferred>,
+        &'static ViewDepthTexture,
+        &'static ViewPrepassTextures,
+        &'static Camera3d,
+        Option<&'static DepthPrepass>,
+        Option<&'static NormalPrepass>,
+        Option<&'static MotionVectorPrepass>,
+    );
+
+    fn run(
+        &self,
+        graph: &mut RenderGraphContext,
+        render_context: &mut RenderContext,
+        (
+            camera,
+            opaque_deferred_phase,
+            alpha_mask_deferred_phase,
+            view_depth_texture,
+            view_prepass_textures,
+            camera_3d,
+            depth_prepass,
+            normal_prepass,
+            motion_vector_prepass,
+        ): QueryItem<Self::ViewQuery>,
+        world: &World,
+    ) -> Result<(), NodeRunError> {
+        let view_entity = graph.view_entity();
+
+        if let Some(msaa) = world.get_resource::<Msaa>() {
+            match msaa {
+                Msaa::Off => (),
+                _ => panic!("MSAA not supported when using deferred rendering."),
+            }
+        }
+
+        let mut color_attachments = vec![];
+        color_attachments.push(
+            view_prepass_textures
+                .normal
+                .as_ref()
+                .map(|view_normals_texture| RenderPassColorAttachment {
+                    view: &view_normals_texture.default_view,
+                    resolve_target: None,
+                    ops: Operations {
+                        load: if normal_prepass.is_some() {
+                            LoadOp::Load // load if the normal_prepass has run
+                        } else {
+                            LoadOp::Clear(Color::BLACK.into())
+                        },
+                        store: true,
+                    },
+                }),
+        );
+        color_attachments.push(view_prepass_textures.motion_vectors.as_ref().map(
+            |view_motion_vectors_texture| RenderPassColorAttachment {
+                view: &view_motion_vectors_texture.default_view,
+                resolve_target: None,
+                ops: Operations {
+                    load: if motion_vector_prepass.is_some() {
+                        LoadOp::Load // load if the motion_vector_prepass has run
+                    } else {
+                        LoadOp::Clear(Color::rgb_linear(0.0, 0.0, 0.0).into())
+                    },
+                    store: true,
+                },
+            },
+        ));
+        color_attachments.push(
+            view_prepass_textures
+                .deferred
+                .as_ref()
+                .map(|deferred_texture| RenderPassColorAttachment {
```

Comment:
Personally I prefer `if let Some(deferred_texture) = view_prepass_textures.deferred` over the maps here.

Reply:
I copied this from the prepass. Looking at it more closely I agree. I didn't even know you could use map this way. I guess this pushes None onto the attachments vec when `view_prepass_textures.deferred` is None? Do you think I should change it in both places for consistency?

Reply:
Hmm, I just tried this and the RenderPipeline expects the None to be here.

Reply:
Tried having it not insert `None` into the RenderPipeline and now it's having issues with the formats not matching in some cases. I'm guessing @IceSentry set it up this way for a reason initially.

Reply:
Iirc it was because there's various combinations of which prepass outputs you can have, and we can't use ifdefs to enumerate every combination easily, so we have to push None for the attachments we skip. We can probably do that better now with ShaderDefVals for the @location() slots...